mne/io/edf/edf.py

```diff
@@ -227,61 +212,61 @@ def _read_segment(self, start=0, stop=None, sel=None, verbose=None,
                         else:
                             warnings.warn('Interpolating stim channel.'
                                           ' Events may jitter.')
-                            oldrange = np.linspace(0, 1, samp + 1,
-                                                   True)
-                            newrange = np.linspace(0, 1, max_samp,
-                                                   False)
+                            oldrange = np.linspace(0, 1, samp + 1, True)
+                            newrange = np.linspace(0, 1, max_samp, False)
                             ch_data = interp1d(
                                 oldrange, np.append(ch_data, 0),
                                 kind='zero')(newrange)
                     elif samp != max_samp:
                         ch_data = resample(x=ch_data, up=max_samp, down=samp,
                                            npad=0)
-                    data[j] = ch_data
-                start_pt = int(max_samp * i)
-                stop_pt = int(start_pt + max_samp)
-                datas[:, start_pt:stop_pt] = data
-        datas *= gains.T
-        datas += offsets
+                    this_data[count, :] = ch_data
+                    count += 1
+                s_off = start - blockstart if bi == 0 else 0
+                e_off = stop_pt - stop + blockstart if bi == n_blk - 1 else 0
+                assert s_off >= 0
+                assert e_off >= 0
+                time_slice = slice(start_pt + (s_off - start + blockstart),
+                                   stop_pt - (e_off + start - blockstart),
+                                   None)
+                data[:, time_slice] = this_data[:, s_off:buffer_size - e_off]
```

Comment:
...what was being done before was simpler, it filled `data` with all channels and timepoints from all blocks, and then cut the data down to size. This made for simpler code, but potentially much higher memory usage, and potentially much worse read efficiency. Optimizations could actually still be done here, e.g. by just using `fid.seek()` to move forward `s_off` number of samples before reading channel data, and then only actually reading `buffer_size - e_off` samples. I'll add that while I'm at it.